sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/healthcare/FhirIO.java

```diff
@@ -0,0 +1,977 @@
+              .apply("File Batches", GroupIntoBatches.ofSize(batchSize))
+              .apply(
+                  ParDo.of(
+                      new ImportFn(fhirStore, tempGcsPath, deadLetterGcsPath, contentStructure)))
+              .setCoder(new HealthcareIOErrorCoder<>(StringUtf8Coder.of()));
+
+      return writeResults
+          .get(Write.FAILED_BODY)
+          .setCoder(new HealthcareIOErrorCoder<>(StringUtf8Coder.of()));
+    }
+
+    static class ImportFn
+        extends DoFn<KV<Integer, Iterable<ResourceId>>, HealthcareIOError<String>> {
+
+      private static final Logger LOG = LoggerFactory.getLogger(ImportFn.class);
+      private final String tempGcsPath;
+      private final String deadLetterGcsPath;
+      private ResourceId tempDir;
+      private final ContentStructure contentStructure;
+      private HealthcareApiClient client;
+      private final String fhirStore;
+
+      ImportFn(
+          String fhirStore,
+          String tempGcsPath,
+          String deadLetterGcsPath,
+          @Nullable ContentStructure contentStructure) {
+        this.fhirStore = fhirStore;
+        this.tempGcsPath = tempGcsPath;
+        this.deadLetterGcsPath = deadLetterGcsPath;
+        if (contentStructure == null) {
+          this.contentStructure = ContentStructure.CONTENT_STRUCTURE_UNSPECIFIED;
+        } else {
+          this.contentStructure = contentStructure;
+        }
+      }
+
+      @Setup
+      public void init() throws IOException {
+        tempDir =
+            FileSystems.matchNewResource(tempGcsPath, true)
+                .resolve(
+                    String.format("tmp-%s", UUID.randomUUID().toString()),
+                    StandardResolveOptions.RESOLVE_DIRECTORY);
+        client = new HttpHealthcareApiClient();
+      }
+
+      /**
+       * Move files to a temporary subdir (to provide common prefix) to execute import with single
+       * GCS URI.
+       */
+      @ProcessElement
+      public void importBatch(ProcessContext context) throws IOException {
+        Iterable<ResourceId> batch = context.element().getValue();
+        List<ResourceId> tempDestinations = new ArrayList<>();
+        List<ResourceId> deadLetterDestinations = new ArrayList<>();
+        assert batch != null;
+        for (ResourceId file : batch) {
+          tempDestinations.add(
+              tempDir.resolve(file.getFilename(), StandardResolveOptions.RESOLVE_FILE));
+          deadLetterDestinations.add(
+              FileSystems.matchNewResource(deadLetterGcsPath, true)
+                  .resolve(file.getFilename(), StandardResolveOptions.RESOLVE_FILE));
+        }
+        FileSystems.rename(ImmutableList.copyOf(batch), tempDestinations);
+        ResourceId importUri = tempDir.resolve("*", StandardResolveOptions.RESOLVE_FILE);
+        try {
+          // Blocking fhirStores.import request.
+          assert contentStructure != null;
+          Operation operation =
```

Comment:
is import method idempotent?

Reply:
@pabloem FYI
Yes. it is idempotent.

> The import operation is idempotent unless the input data contains multiple valid resources with the same ID but different contents. In that case, after the import completes, the store contains exactly one resource with each ID, but the duplicate entries could contain any version of the contents. For example, importing a million resources with the same ID writes only one resource to the store.
[FHIR Import Public Docs](https://cloud.google.com/healthcare/docs/concepts/fhir-import)